### ISSUE TITLE
API to persist column widths

### ISSUE THREAD

**Issue by seancolsen** [mathesar-foundation/mathesar#1421]
## Current behavior

- The user can resize columns on the front end, but the sizes do not persist after the user closes the table tab within Mathesar.

## Desired behavior

- Column widths are saved via the API. We will need to design the API for this.

## Scope

- This ticket covers the API changes only.
- Front end changes are tracked via #3263

**Comment by seancolsen** [comment]
@mathemancer @dmos62 @silentninja I am working on this as part of my [Usability Improvements](https://wiki.mathesar.org/en/projects/2023-04-usability-improvements) project. I predict this will be a simple enough back-end change that I can make it myself. But I would like to know if any of you have opinions about the API design.

What should the request look like to set a column width?

Some examples:

- `PATCH /api/db/v0/tables/{table_id}/columns/{column_id}/`

    ```json
    {
      "display_options": {
        "width":100
      }
    }
    ```

- `PATCH /api/db/v0/tables/{table_id}/columns/{column_id}/`

    ```json
    {
      "width":100
    }
    ```

I can think of others too. We could potentially use the `ui` namespace too. I don't have many opinions about this. Do any of you?

**Comment by dmos62** [comment]
> I'd rather avoid sending the entire display options blob along with the column width because then we run the risk of overwriting display options in the case where (for example), Alice loads the display options into the front end, Bob writes a display option, and then Alice adjusts a column width (thereby overwriting Bob's change).

Makes sense. Won't get us to what users expect out of concurrent editing these days, but it's something.

**Comment by silentninja** [comment]
Having a `partial` update for `display_options` leads to a lot of validation-related problems. For example, in the case of a `boolean` type column which has a `null` value for the `display_options`, the user can send a request to only update the `width` field and it would end up creating a `display_options` object without the mandatory `input` field. 

Implementing the `merge` option for this API is quite easy but it would lead to bugs that are hard to find. 

The actual problem is real-time concurrency for global `display_options` and I don't think having a `merge` feature would solve it effectively. I would rather prefer to solve this problem by having **user-specific** `display_options` which are inherited from the **global** `display_options` which totally avoids concurrency-related issues.

**Comment by seancolsen** [comment]
@silentninja

> having user-specific display_options which are inherited from the global display_options

I'm having a hard time imagining how this would work without a lot of complexity. Perhaps you could elaborate? You want user-specific display_options? Why? This is not an implementation detail &mdash; it's a different product design.

I dug up the previous discussion about this. See [this email](https://groups.google.com/a/mathesar.org/g/mathesar-developers/c/uGM_Dwa8UEM/m/5gwrlf3kAwAJ) where I make a case for global display options.

Notice that Kriti said:

> Okay, I'm convinced – let's start with global display options and figure out if we need per-user options based on user feedback.

---

You said:

> it would end up creating a `display_options` object without the mandatory `input` field.

That's not how I imagine the logic working. I see it like this:

The validator looks at three `display_options` dictionaries, taking entries from them with precedence defined by the following order.

1. The request parameters (highest precedence)
1. The existing values
1. The default values known for the type (lowest precedence)

In your boolean example, the `input` field would come from that third dictionary, the default values. What bugs would this logic lead to?

**Comment by silentninja** [comment]
> I dug up the previous discussion about this. See [this email](https://groups.google.com/a/mathesar.org/g/mathesar-developers/c/uGM_Dwa8UEM/m/5gwrlf3kAwAJ) where I make a case for global display options.

The apps you mentioned(like Google sheets) in the email reflect changes made to the `display_options` to other users in real-time. We are missing that feature in Mathesar and hence my suggestion is to use a user-specific `display_option` in addition to global `display_options`.

> I'm having a hard time imagining how this would work without a lot of complexity. Perhaps you could elaborate? You want user-specific display_options? Why? This is not an implementation detail — it's a different product design.

I agree it is a different product design. I am suggesting we should take this route if we plan on fixing concurrency issues as it prevents concurrency issues altogether by limiting to just one user instead of having to reflect the display_option changes in real time.


`display_options` is a single JSON field. So a `PATCH` request is made for the `Column` object and not the `display_options` field. So the entire JSON object is replaced with the new JSON object(a partial JSON object in this case) sent in the request.

> The default values known for the type (lowest precedence)

This step is another problem with the proposed solution. Columns do not have a default `display_options` object because, when a column type changes in the database, we validate the `display_options` by checking for the existence of mandatory fields in the JSON object. If the validation fails (when the column type and the `display_options` do not match), we reset the `display_options` to `null`. 

If we were to set default values for `display_options`, we would not be able to validate them because all the values will be considered optional. So, when a column type changes in the database, we would be unable to reset the `display_options`, resulting in `display_options` containing extraneous values.

If I change the `NUMERIC` column to `BOOLEAN` type. The `display_options` will be
{percentage": True, "input": "checkbox"} instead of {"input": "checkbox"}

`"percentage": True` gets carried over from the `NUMERIC` type


To conclude, I can make this change but there are too many hacks I need to implement which I am afraid **might** lead to some bugs, so just considering the tradeoffs at this point.

**Comment by seancolsen** [comment]
@silentninja

- I don't think realtime collaboration is especially relevant here. Yes, it would make make data races _less likely_ but it would not eliminate them. Even with realtime collaboration, I don't think we'd want to design an API which writes unrelated display_options data when a users resize a column.

- This conversation highlights some tension between product design (where I want global display_options), and implementation feasibility (where you want per-user display_options). Obviously product design affects implementation. And generally speaking, I also think implementation challenges should affect product design too. If a feature is just too hard to implement, then yes, we should reconsider the product design. But the bar for that sort of reconsideration should be quite high. We should be creative and dogged and make sure we've exhausted all our options before reconsidering the product design. I don't think we've met that bar in this case.

- What about this previous [suggestion](https://github.com/centerofci/mathesar/issues/1421#issuecomment-1539947235) of mine?

    > `PATCH /api/db/v0/tables/{table_id}/columns/{column_id}/`
    > 
    > ```json
    > {
    >   "width": 100
    > }
    > ```

    With this API design, the `width` property would live alongside display_options. Would that be okay?

**Comment by silentninja** [comment]
> What about this previous https://github.com/centerofci/mathesar/issues/1421#issuecomment-1539947235 of mine?

I don't think it should belong there the `width` field is not part of the `Column` object, instead it is part of the display settings used by the UI.


Before I propose a different solution, I would like to highlight the key issues causing the friction

- The `display_options` is a `JSON` field on the backend. So the backend considers it to be a `property` of the `Column` and not a [restful `resource`](https://cloud.google.com/apis/design/resources).  So when the client sends a `PATCH`  request to the `/columns/{id}/`, the backend treats it as a `PATCH` only for the `Column` resource and not the sub-resources or fields within the `Column`. We decided to follow this pattern [when we implemented `display_options` ](https://github.com/centerofci/mathesar/issues/390#issuecomment-971913896)
- Setting the `display_options` to `null` leads to ambiguous state. 

> I don't think realtime collaboration is especially relevant here.

It is relevant for our discussion. If we didn't have concurrency issues, the frontend can just send the complete `display_options` object when updating without the need for `merge` feature.

#### Proposed Solutions:
- Even if we don't implement `merge` for the `display_options`, it should always contain default values so there is one place of responsibility for the `display_options` object creation. I think backend can take up the responsibility because we also need to change the `display_options` when the column type changes.
- Ignore the need for concurrency and send the full body of the `display_options`.
- (OR) Make changes to the backend to `merge` display_options
- (OR) Add support for user specific `display_options`


I think it would be helpful to question "Do we need to support concurrency?" to come up with solution

**Comment by silentninja** [comment]
I had a call with @seancolsen and we finalized on some key decisions but we still cannot come to a conclusion. @kgodey we would like to hear your opinion on the issue.

I will be using the following example `Column` object as reference in various places

```
{
   'id': 1,
   'type': '
   'display_options': {'show_as_percentage': True, 'input': 'checkbox'}
}
```

#### 1. Make changes to the backend to allow for merge/partial_update to a `display_options` 

- Con: There is a bug with [DRF](https://github.com/centerofci/mathesar/issues/390#issuecomment-971913896) where it would partially validate the nested object but will end up saving the partial value as the complete object.
A partial_update request with `'display_options': {'width': 5}` to `/api/columns/1/` will end up with 
`'display_options': {'width': 5}` instead of `'display_options': {'show_as_percentage': True, 'width': 5, 'input': 'checkbox'}`
- Con: Fixing the bug would mean overriding some of `DRF` functions (`save` and `validate`)  which could lead to more code and potentially more bugs. 
- Pro: There won't be weird UX due to concurrency issues 
- Argument against the "Pros": The concurrency issue is a problem of the distant future.
- Argument against the "Cons": Technical implementation should follow product and look out for what is convinent for the users.


#### 2. Ignore the need for concurrency and send the full body of the `display_options` field.

- Con:  We could end with concurrency issues. If a "User A" is making changes to the  `width` display_option and  "User B" is setting `show_percentage` display_options as `False`. We could end up with a race condition. We have a higher chance of this occuring in our app as we don't reflect changes in realtime. So if a user stays on the `Column` page for year without refreshing, he could end up reverting all the changes made to that column `display_options` by other users.
-  Pro: We follow what we have been doing so far, so no code changes to make
- Argument against the "Pros": `width` display_option is odd one. With other display_options you can see what will be affect on the backend as the UI contains all the `display_options` to be sent in the actions pane but when changing the `width` the user won't realise what is being affected.
- Argument against the "Cons": We can fix this when a user encounters this issue as the data that is lost isn't mission critical.

#### 3. Move the `width` field out of `display_options` into the `Column` object

- Pro:  We can send a `PATCH` request to `/columns/` avoiding all the cons of previous suggestions.
- Pro: We are making the codebase messy but makes the UX better.
- Cons: From a technical standpoint `width` is not a property of the `Column` object. So it will be confusing for anyone looking at the code to undertand where this property is coming from or its relation to the `Column` object. 
- Cons: What if we end up with more `display_options` in the similar vein down the line, it could end up creating a mess
- Dirty fix but gets the job done



@kgodey Do you think we could be affected by the concurrency issues in this case? 

I prefer to go with (2) as I find don't find the concurrency issues to be a pressing problem and we don't have enough data to know if it is a problem.

**Comment by seancolsen** [comment]
> 3. ... We are making the codebase messy but making the UX better.

This is why I prefer (3) and why we had a hard time reaching an agreement on the call.

Mukesh wants (2). I want (3). We'd like opinions from others to help break the tie. @dmos62 @mathemancer would you care to weigh in?

**Comment by kgodey** [comment]
Do all the current display options affect the content of cells (rather than the column itself?) If so, I'd consider renaming it to `cell_display_options` and going with (3). I would name it `ui_width` or something more clear than `width` though, so that it's clear it applies to the UI and isn't something DB related.

I also think we should move this API to the `ui` namespace instead of the `db` namespace, since it contains UI specific elements already. I think it will help us care less about the integrity of the `Column` object if it's in the UI namespace.

Once it's in the UI namespace, we could also unpack all of the other `display_options` into top level keys, but I don't think we need to do that now.

**Comment by silentninja** [comment]
I would like to propose a different solution which I think would solve both the problem. I would like to introduce a separate resource called `column_settings` which would be implemented in the same way as `table_settings`. This `column_settings`  will contain properties like `width`.  So instead of a `JSON` field like `display_options`, this will be a separate resource linked by a `One-to-One` relation. 

So the `tables/{id}/columns/` respone would look like
```
[
    {
       'id': 1,
       'type': '
       'display_options': {'show_as_percentage': True, 'input': 'checkbox'},
       'settings': {'id': 1, 'width': 20}
    }
]
```

This will prevent concurrency issues as you can send a `PATCH` request to `/tables/{table_id}/columns/{column_id}/settings/{settings_id}/`.


The advantages are the same as for `table_settings`
1. The `settings` object is decoupled from `Column` objects which means in future we can add support for different devices or have user-specific settings and attach the `settings` to the `Column` object dynamically when returning back the API.

The disadvantages are
1. Client would get only the `settings` object when sending an update to `/columns/settings/id` which would mean they would have to update the `settings` object in the `Column` object they already have. 
2. Additional queries needed - We already have the infrastructure to avoid N+1 queries if needed, so this shouldn't be a problem
3. Nested routes - The route can be just called `/column_settings/`. Having more than 2 levels is something to be concerned about from an API but from an implementation perspective, I don't see any problem. 


@seancolsen and @kgodey Are you okay with this proposal?

**Comment by silentninja** [comment]
> I think we should move this API to the ui namespace instead of the db namespace, since it contains UI specific elements already. It will help us care less about the integrity of the Column object if it's in the UI namespace.


Can you elaborate on this? Do you mean having a `...ui/tables/{table_id}/columns/{id}` API which will only allow updates for the  `width` field?


Also, I am not concerned about the API, I am more concerned with how tightly coupled the `Column` model will end up. The more configuration we offer to the users, the more we might have to shove into the `Column` object. Also, this property is UI-specific which means we might have to support dynamic values based on different UI and maybe different users. 


On a side note, I don't want to drag this conversation further and steer it in a different direction but I think this `display_option` should be user-specific. Why would someone using Mathesar on a high resolution be okay with the settings set by a user viewing Mathesar on a low-resolution monitor? This question is a different conversation on its own but just putting it out for visiblity.

**Comment by seancolsen** [comment]
@silentninja

With the `settings` approach, let's pretend that in several months we decide to add a new setting called `conditional_formatting`. Would the front end be able to update the `width` value without making any changes to `conditional_formatting`? That is, can we implement `merge` for the settings?

**Comment by dmos62** [comment]
I'm gravitating towards either Mukesh's [last proposal](https://github.com/centerofci/mathesar/issues/1421#issuecomment-1610292752) or 1. I think of `display_options` as a blackbox JSON blob that the backend is storing for the frontend. The more explicit this is, the more I like it. I would really like to not to explode it into the containing model (which is what 3 is proposing). Unless, `display_options` isn't a blackbox (e.g. we want to validate it somehow).

**Comment by kgodey** [comment]
I'm fine with this approach. @silentninja I don't know if you still want me to elaborate on what I was saying before, but it seems irrelevant now.

**Comment by silentninja** [comment]
> I don't know if you still want me to elaborate on what I was saying before, but it seems irrelevant now.

It will be great if you can, it might be useful in the future when evaluating the options we have.

**Comment by kgodey** [comment]
> Can you elaborate on this? Do you mean having a `...ui/tables/{table_id}/columns/{id}` API which will only allow updates for the `width` field?

No, I meant moving the entire `...db/tables/{table_id}/columns/{id}` to `...ui/tables/{table_id}/columns/{id}`. Currently we're not aiming for our API to be broadly usable beyond the UI and we have `display_options` in the API, which is only UI related (it doesn't have anything to do with the DB).

The idea with the DB namespace is that it should only have data that maps directly to the DB and the UI namespace is tailored to whatever makes the UI function well (which can be a superset of what's available in the APIs in the DB namespace, or something totally different). UI-specific things like display options and column width shouldn't be in the DB namespace, but DB specific data like column name etc. can be in the UI namespace if we want to reduce the number of requests the UI needs to make.

This is the theory, but we haven't actually implemented this in practice, and I think we should eventually (probably before we freeze the APIs for v1.0).

**Comment by silentninja** [comment]
@seancolsen I added `Column Settings` API and updated the issue descriptions to include the newly introduced API.
